### ISSUE TITLE
Hide hot restart and hot reload buttons based on the availability of the service

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/devtools#1941]
In https://github.com/flutter/devtools/pull/1930, we hid these buttons for non-flutter apps, but this is problematic for dart server apps that support hot reload. 

We need to hide the buttons and update the available actions space on the app bar when the hot reload and hot restart services are not available

**Comment by kenzieschmoll** [comment]
@julemand101 do you have a small sample app (non-flutter) that has the hot reload service available? I'm unable to repro this behavior with the Dart CLI apps I am using.

**Comment by kenzieschmoll** [comment]
After further investigation, it doesn't appear that the `reloadServices` service is registered for Dart CLI apps by default. The Flutter framework calls `vmService.registerService('reloadSources')`, which is how DevTools is aware of the service. Calling that service ultimately bubbles down to calling the method `VmService.reloadSources`. The `VmService.reloadSources` RPC is what other packages like `hotreloader` also use: https://github.com/vegardit/dart-hotreloader/blob/main/lib/hotreloader.dart#L255. 

@bkonyi what is the best way to know if the `reloadSources` functionality is available for a Dart CLI app?

**Comment by julemand101** [comment]
@kenzieschmoll Well, one case I know of is just running the program with JIT and the Dart VM Service enabled since that opens up for any program to ask the Dart program to reload its sources.

A simple program like this can be used (where we later will remove the comment):
```dart
import 'dart:async';

void main() {
  Timer.periodic(Duration(seconds: 3), myFunction);
}

void myFunction(Timer _) {
  // print('Hello world');
}
```

1. Run Dart code with `dart --observe --serve-observatory program.dart`
2. Visit the old Dart Observatory where a link to it is printed as the first line of output. E.g.:
```
The Dart VM service is listening on http://127.0.0.1:8181/Fdd5tkhH8Ds=/
```
3. Click on the main isolate and you can then find this panel:
![image](https://github.com/flutter/devtools/assets/116227/0a1a15b2-9f5d-450d-abcc-df76b7cee9fa)

4. Remove the `//` in the source code and click "Reload Source".
5. Notice the program now starts printing the message.

A more advanced example can be made where we can use `dev.Service.controlWebServer(enable: true)` from `dart:developer` to open the Dart VM Service and then connect to that and trigger the reloads.

**Comment by kenzieschmoll** [comment]
In the linked PR https://github.com/flutter/devtools/pull/6341, I have switched to calling `VmService.reloadSources` for non-flutter apps (like observatory does https://github.com/dart-lang/sdk/blob/1c6c8ef94d7d082853f34d930cad928bc57bebac/runtime/observatory/lib/src/repositories/isolate.dart#L28), and I'm seeing something strange. 

@bkonyi interested to hear if you have any ideas why this might be.

In these two scenarios, I see the following results after running `dart run --serve-observatory --observe`:
1. Open Observatory, navigate to the main isolate and click "Reload Source". This succeeds. Then open DevTools (running with the changes in the linked PR above), and click the Hot Reload button. This succeeds.
2. Open DevTools first and click Hot Reload. This fails with the error below.

So Observatory must be doing something on its initial open that prevents this error from getting thrown.

Error: 
```
../../runtime/vm/isolate_reload.cc: 671: error: An error occurred while rejecting the most recent compilation results: Bad state: Incremental compiler created for expression compilation only; cannot reject
#0      IncrementalCompiler.reject (package:vm/incremental_compiler.dart:169:7)
#1      IncrementalCompilerWrapper.reject (file:///opt/s/w/ir/x/w/sdk/pkg/vm/bin/kernel_service.dart:351:45)
#2      _processLoadRequest (file:///opt/s/w/ir/x/w/sdk/pkg/vm/bin/kernel_service.dart:843:25)
#3      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)

version=3.2.0-122.0.dev (dev) (Tue Aug 29 13:03:15 2023 -0700) on "macos_x64"
pid=36187, thread=12035, isolate_group=main(0x7febe2045a00), isolate=main(0x7febe2019c00)
os=macos, arch=x64, comp=no, sim=no
isolate_instructions=10952fa00, vm_instructions=10952fa00
fp=70000d0b0d90, sp=70000d0b0d58, pc=109798064
  pc 0x0000000109798064 fp 0x000070000d0b0d90 dart::Profiler::DumpStackTrace(void*)+0x64
  pc 0x000000010952fc14 fp 0x000070000d0b0e70 dart::Assert::Fail(char const*, ...) const+0x84
  pc 0x00000001096cf1bb fp 0x000070000d0b1110 dart::IsolateGroupReloadContext::Reload(bool, char const*, char const*, unsigned char const*, long)+0x133b
  pc 0x00000001096c777c fp 0x000070000d0b11d0 dart::IsolateGroup::ReloadSources(dart::JSONStream*, bool, char const*, char const*, bool)+0x1ec
  pc 0x00000001097fca59 fp 0x000070000d0b1210 dart::Service::PrintJSONForVM(dart::JSONStream*, bool)+0x7d29
  pc 0x00000001097f36e5 fp 0x000070000d0b1820 dart::Service::InvokeMethod(dart::Isolate*, dart::Array const&, bool)+0x345
  pc 0x00000001097f3db1 fp 0x000070000d0b1840 dart::Service::HandleIsolateMessage(dart::Isolate*, dart::Array const&)+0x11
  pc 0x00000001096c56dd fp 0x000070000d0b1de0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message>>)+0x21d
  pc 0x00000001096ed847 fp 0x000070000d0b1e50 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x127
  pc 0x00000001096edeec fp 0x000070000d0b1ea0 dart::MessageHandler::TaskCallback()+0x1ec
  pc 0x000000010981db3b fp 0x000070000d0b1f20 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x14b
  pc 0x000000010981df99 fp 0x000070000d0b1f50 dart::ThreadPool::Worker::Main(unsigned long)+0x79
  pc 0x000000010979458e fp 0x000070000d0b1fb0 dart::OSThread::GetMaxStackSize()+0xde
  pc 0x00007ff809e731d3 fp 0x000070000d0b1fd0 _pthread_start+0x7d
  pc 0x00007ff809e6ebd3 fp 0x000070000d0b1ff0 thread_start+0xf
-- End of DumpStackTrace
[1]    36187 abort      dart run --serve-observatory --observe
```

**Comment by bkonyi** [comment]
As discussed offline, this is due to `dart run --observe` performing compilation from source to kernel before actually running the script under `bin/`. Running the script directly (e.g., `dart run --observe bin/foo.dart`) should work correctly.

The above behavior can also be reproduced if you explicitly compile the source to kernel, running the kernel directly, and then trying to call `reloadSources`:

```
$ dart compile kernel bin/foo.dart
Compiling bin/foo.dart to kernel file bin/foo.dill.
$ dart run --observe bin/foo.dill
The Dart VM service is listening on http://127.0.0.1:8181/Xn25WvsnzXs=/
The Dart DevTools debugger and profiler is available at: http://127.0.0.1:8181/Xn25WvsnzXs=/devtools?uri=ws://127.0.0.1:8181/Xn25WvsnzXs=/ws
```

Calling `reloadSources` will then give the same error:

```
../../runtime/vm/isolate_reload.cc: 671: error: An error occurred while rejecting the most recent compilation results: Bad state: Incremental compiler created for expression compilation only; cannot reject
#0      IncrementalCompiler.reject (package:vm/incremental_compiler.dart:169:7)
#1      IncrementalCompilerWrapper.reject (file:///opt/s/w/ir/x/w/sdk/pkg/vm/bin/kernel_service.dart:351:45)
#2      _processLoadRequest (file:///opt/s/w/ir/x/w/sdk/pkg/vm/bin/kernel_service.dart:843:25)
#3      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)

...
```
